File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/NumberStringConverterForFloatingPointNumbers.java
package com.cedarsoft.commons.javafx;

import java.text.NumberFormat;
import java.util.Locale;

import javax.annotation.Nonnull;

import javafx.util.converter.NumberStringConverter;

/**
 * A {@link NumberStringConverter} for floating point values only.
 *
 * @author Christian Erbelding (<a href="mailto:ce@cedarsoft.com">ce@cedarsoft.com</a>)
 */
public class NumberStringConverterForFloatingPointNumbers extends NumberStringConverter {
  @Nonnull
  private final NumberFormat doubleNumberFormat;

  public NumberStringConverterForFloatingPointNumbers() {
    this(Locale.getDefault());
  }

  public NumberStringConverterForFloatingPointNumbers(@Nonnull Locale locale) {
    this(NumberFormat.getNumberInstance(locale));
  }

  public NumberStringConverterForFloatingPointNumbers(@Nonnull NumberFormat numberFormat) {
    doubleNumberFormat = numberFormat;
  }

  @Nonnull
  @Override
  protected NumberFormat getNumberFormat() {
    return doubleNumberFormat;
  }

  @Override
  public Number fromString(String value) {
    if ("-".equals(value)) {
      return 0.0;
    }
    return super.fromString(value);
  }

}

File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/NumberStringConverterForFloatingPointNumbers.kt
package com.cedarsoft.commons.javafx

import javafx.util.converter.NumberStringConverter
import java.text.NumberFormat
import java.util.Locale

/**
 * A [NumberStringConverter] for floating point values only.
 *
 * @author Christian Erbelding ([ce@cedarsoft.com](mailto:ce@cedarsoft.com))
 */
class NumberStringConverterForFloatingPointNumbers(
  private val doubleNumberFormat: () -> NumberFormat
) : NumberStringConverter() {

  constructor(doubleNumberFormat: NumberFormat) : this({ doubleNumberFormat })

  @JvmOverloads
  constructor(locale: Locale = Locale.getDefault()) : this(NumberFormat.getNumberInstance(locale))

  override fun getNumberFormat(): NumberFormat {
    return doubleNumberFormat()
  }

  override fun fromString(value: String): Number {
    return if ("-" == value) {
      0.0
    } else {
      super.fromString(value)
    }
  }
}
